File: apps/web/src/types/events.ts
export interface EventSummary {
  id: string;
  name: string;
  slug: string;
  kind: "football" | "custom";
  status: "draft" | "active" | "finished";
  endsAt: string | null;
  isHistorical: boolean;
  coverUrl?: string | null;
  coverAssetUrl?: string | null;
  externalUrl?: string | null;
}
export interface EventRecord {
  id: string;
  name: string;
  slug: string;
  kind: "football" | "custom";
  origin: "system" | "user";
  status: "draft" | "active" | "finished";
  createdBy: string | null;
  startsAt: string | null;
  endsAt: string | null;
  createdAt: string;
  updatedAt: string;
  description?: string | null;
  coverUrl?: string | null;
  coverAssetUrl?: string | null;
  externalUrl?: string | null;
  poolCreationEnabled?: boolean;
  currentPublishedVersionId?: string | null;
  archivedAt?: string | null;
}
export interface EventVersionHistory {
  id: string;
  versionNumber: number;
  state: "working" | "published";
  isCurrentPublished: boolean;
  name: string;
  fingerprint: string;
  baseFingerprint: string | null;
  createdAt: string;
  updatedAt: string;
  itemCount: number;
  optionCount: number;
  poolCount: number;
}
export type ManifestAction = "create" | "noChange" | "safeUpdate" | "conflict" | "rejected";
export interface ManifestDiffEntry { category: string; path: string; change: string; }
export interface ManifestPreview {
  action: ManifestAction;
  name: string;
  slug: string;
  schemaVersion: number;
  itemCount: number;
  optionCount: number;
  linkCount: number;
  manifestFingerprint: string;
  baseFingerprint: string;
  safeChanges: ManifestDiffEntry[];
  blockedChanges: ManifestDiffEntry[];
}
export interface ManifestApplyResult {
  action: ManifestAction;
  eventId: string | null;
  itemCount: number;
  optionCount: number;
  linkCount: number;
  versionId: string | null;
  state: "working" | "published";
}
export interface CustomQuestionOption {
  id: string;
  label: string;
  sortOrder: number;
  imageUrl?: string | null;
  imageAssetUrl?: string | null;
  links?: OptionLink[];
  mediaSeen?: boolean;
}
export interface OptionLink { kind: "video" | "audio" | "official" | "other"; label: string; url: string; sortOrder: number; }
export interface EventShowcase { name: string; description: string | null; coverUrl: string | null; coverAssetUrl?: string | null; externalUrl: string | null; startsAt: string | null; endsAt: string | null; itemCount: number; answeredCount: number; isHistorical: boolean; }
export interface AssetResponse {
  assetId: string;
  sha256: string;
  mediaType: string;
  width: number;
  height: number;
  byteSize: number;
  url: string;
  variants: Record<string, string>;
}
export interface PackagePreview {
  manifest: ManifestPreview;
  assetCount: number;
  existingAssetCount: number;
  addedAssetCount: number;
}
export interface PackageApplyResult {
  result: ManifestApplyResult;
  assetCount: number;
  addedAssetCount: number;
}

export interface PackageExportPreview {
  assetCount: number;
  externalImageCount: number;
  externalImages: { question: string; optionLabel: string | null; url: string }[];
}
export interface CustomQuestion {
  itemId: string;
  kind: "single_choice" | "numeric" | "multiple_choice";
  title: string;
  lockAt: string;
  revealAt: string;
  sortOrder: number;
  status: "draft" | "open" | "locked" | "resolved";
  currentOptionId: string | null;
  correctOptionId: string | null;
  correctPoints: number;
  incorrectPoints: number;
  options: CustomQuestionOption[];
  decimalPlaces?: number;
  unitLabel?: string | null;
  minValue?: string | null;
  maxValue?: string | null;
  currentValue?: string | null;
  resultValue?: string | null;
  resultStatus?: "resolved" | "not_representable" | "pending_decision" | null;
  exactPoints?: number;
  tolerance?: string;
  withinTolerancePoints?: number;
  minSelections?: number;
  maxSelections?: number | null;
  currentOptionIds?: string[];
  correctOptionIds?: string[];
  partialPoints?: number;
}
export interface CustomMemberPredictions { userId: string; username: string; predictions: { itemId: string; title: string; optionLabel: string; points: number | null }[]; }

File: apps/server/src/api/dto/events.rs
use serde::Deserialize;

#[derive(Deserialize)]
#[serde(rename_all = "camelCase")]
pub(crate) struct CreatePoolBody {
    pub(crate) name: String,
    #[serde(default)]
    pub(crate) event_id: Option<String>,
}
#[derive(Deserialize)]
#[serde(rename_all = "camelCase")]
pub(crate) struct CreateEventBody {
    pub(crate) name: String,
    pub(crate) starts_at: Option<String>,
    pub(crate) ends_at: Option<String>,
}
#[derive(Deserialize)]
#[serde(rename_all = "camelCase")]
pub(crate) struct CreateItemBody {
    pub(crate) title: String,
    pub(crate) lock_at: String,
    pub(crate) reveal_at: String,
}
#[derive(Deserialize)]
#[serde(rename_all = "camelCase")]
pub(crate) struct CreateNumericItemBody {
    pub(crate) title: String,
    pub(crate) lock_at: String,
    pub(crate) reveal_at: String,
    pub(crate) decimal_places: i64,
    pub(crate) unit_label: Option<String>,
    pub(crate) min_value: Option<String>,
    pub(crate) max_value: Option<String>,
}
#[derive(Deserialize)]
#[serde(rename_all = "camelCase")]
pub(crate) struct CreateMultipleChoiceItemBody {
    pub(crate) title: String,
    pub(crate) lock_at: String,
    pub(crate) reveal_at: String,
    pub(crate) min_selections: i64,
    pub(crate) max_selections: Option<i64>,
}
#[derive(Deserialize)]
pub(crate) struct CreateOptionBody {
    pub(crate) label: String,
}
#[derive(Deserialize)]
#[serde(rename_all = "camelCase")]
pub(crate) struct UpdateOptionMediaBody {
    pub(crate) image_url: Option<String>,
    #[serde(default)]
    pub(crate) links: Vec<crate::custom_events::BuilderOptionLink>,
}
#[derive(Deserialize)]
#[serde(rename_all = "camelCase")]
pub(crate) struct UpdateEventBody {
    pub(crate) name: String,
    pub(crate) starts_at: Option<String>,
    pub(crate) ends_at: Option<String>,
    pub(crate) description: Option<String>,
    pub(crate) cover_url: Option<String>,
    pub(crate) external_url: Option<String>,
}
#[derive(Deserialize)]
#[serde(rename_all = "camelCase")]
pub(crate) struct UpdateItemBody {
    pub(crate) title: String,
    pub(crate) lock_at: String,
    pub(crate) reveal_at: String,
}
#[derive(Deserialize)]
pub(crate) struct MoveBody {
    pub(crate) direction: i64,
}
#[derive(Deserialize)]
#[serde(rename_all = "camelCase")]
pub(crate) struct ManifestPreviewBody {
    pub(crate) content: String,
    #[serde(default)]
    pub(crate) filename: Option<String>,
}
#[derive(Deserialize)]
#[serde(rename_all = "camelCase")]
pub(crate) struct ManifestApplyBody {
    pub(crate) content: String,
    pub(crate) base_fingerprint: String,
    #[serde(default)]
    pub(crate) filename: Option<String>,
}
#[derive(Deserialize)]
#[serde(rename_all = "camelCase")]
pub(crate) struct EventAvailabilityBody {
    pub(crate) enabled: bool,
}
